test/functional/passwords_controller_test.rb

```diff
@@ -96,6 +96,84 @@ class PasswordsControllerTest < ActionController::TestCase
     end
   end
 
+  context "on POST to webauthn_edit" do
+    context "with webauthn enabled" do
+      setup do
+        @user = create(:user)
+        @webauthn_credential = create(:webauthn_credential, user: @user)
+        get :edit, params: { token: @user.confirmation_token, user_id: @user.id }
+        @challenge = session[:webauthn_authentication]["challenge"]
+        @origin = "http://localhost:3000"
+        @rp_id = URI.parse(@origin).host
+        @client = WebAuthn::FakeClient.new(@origin, encoding: false)
+        WebauthnHelpers.create_credential(
+          webauthn_credential: @webauthn_credential,
+          client: @client
+        )
+        post(
+          :webauthn_edit,
+          params: { user_id: @user.id,
+            token: @user.confirmation_token,
+            credentials:
+            WebauthnHelpers.get_result(
+              client: @client,
+              challenge: @challenge
+            ) }
+        )
+      end
+
+      should respond_with :success
+      should "display edit form" do
+        assert page.has_content?("Reset password")
+      end
+    end
+
+    context "when not providing credentials" do
+      setup do
+        @user = create(:user)
+        @webauthn_credential = create(:webauthn_credential, user: @user)
+        get :edit, params: { token: @user.confirmation_token, user_id: @user.id }
+        post :webauthn_edit, params: { user_id: @user.id, token: @user.confirmation_token }, format: :html
+      end
+
+      should respond_with :unauthorized
+      should "set flash notice" do
+        assert_equal "Credentials required", flash[:alert]
+      end
```

Comment:
I'm confused about how this test works, could you explain it to me? It seems like the assertion says that creds aren't provided, but in the `setup` it looks to me like they do? I'm sure it works, I am just curious about what I am overlooking.

Reply:
Yeah in this post, no credentials are passed in.

In the 'happy path' test, the post looks like this:
```ruby
post(
  :webauthn_edit,
  params: { user_id: @user.id,
    token: @user.confirmation_token,
    credentials:
    WebauthnHelpers.get_result(
      client: @client,
      challenge: @challenge
    ) }
)
 ```
 
 But, in this test, the post looks like:
 ```ruby
 post :webauthn_edit, params: { user_id: @user.id, token: @user.confirmation_token }, format: :html
 ```
 
So, the `credentials` field isn't passed into the params, which contains the WebAuthN client and the challenge. So, we expect a response of `unauthorized` because none of those credential params were passed through. Does that answer your question?

Reply:
This was helpful, thankyou.